File: xmlBlaster/demo/javaclients/HelloWorld2.java
// xmlBlaster/demo/javaclients/HelloWorld2.java
import org.xmlBlaster.util.Log;
import org.xmlBlaster.util.ConnectQos;
import org.xmlBlaster.util.DisconnectQos;
import org.xmlBlaster.client.I_Callback;
import org.xmlBlaster.client.UpdateKey;
import org.xmlBlaster.client.UpdateQos;
import org.xmlBlaster.client.protocol.XmlBlasterConnection;
import org.xmlBlaster.engine.helper.MessageUnit;


/**
 * This client connects to xmlBlaster and subscribes to a message. 
 * <p />
 * We then publish the message and receive it asynchronous in the update() method. 
 * <p />
 * Invoke: java HelloWorld2
 * @see <a href="http://www.xmlBlaster.org/xmlBlaster/doc/requirements/interface.html" target="others">xmlBlaster interface</a>
 */
public class HelloWorld2 implements I_Callback
{
   public HelloWorld2(String[] args) {
      try {
         XmlBlasterConnection con = new XmlBlasterConnection(args);

         ConnectQos qos = new ConnectQos(null, "joe", "secret");
         con.connect(qos, this);  // Login to xmlBlaster, register for updates

         con.subscribe("<key oid='HelloWorld2'/>", "<qos/>");

         con.publish(new MessageUnit("<key oid='HelloWorld2'/>", "Hi".getBytes(),
                                     "<qos/>"));

         try {
            Thread.currentThread().sleep(1000); // wait a second
            Log.info("", "Hit a key to logout and terminate ...");
            System.in.read();
         } catch(Exception e) { }

         con.erase("<key oid='HelloWorld2'/>", null);
         con.disconnect(null);
      }
      catch (Exception e) {
         Log.panic("", e.toString());
      }
   }

   public String update(String cbSessionId, UpdateKey updateKey, byte[] content,
                        UpdateQos updateQos)
   {
      Log.info("HelloWorld2", "Received asynchronous message '" +
         updateKey.getOid() + "' state=" + updateQos.getState() + " from xmlBlaster");
      return "";
   }

   public static void main(String args[]) {
      new HelloWorld2(args);
   }
}

File: xmlBlaster/demo/c++/HelloWorld2.cpp

#include <util/xmlBlasterDef.h>
#include <client/XmlBlasterAccess.h>
#include <util/qos/ConnectQos.h>
#include <util/XmlBlasterException.h>
#include <util/Global.h>
#include <util/Log.h>
#include <client/I_Callback.h>
#include <client/UpdateKey.h>
#include <client/UpdateQos.h>
#include <util/MessageUnit.h>
#include <util/PlatformUtils.hpp>
#include <util/Timestamp.h>

#include <util/qos/PublishQos.h>

/**
 * This client connects to xmlBlaster and subscribes to a message.
 * <p />
 * We then publish the message and receive it asynchronous in the update() method.
 * <p />
 * Invoke: HelloWorld2
 * @see <a href="http://www.xmlBlaster.org/xmlBlaster/doc/requirements/interface.html" target="others">xmlBlaster interface</a>
 */

using namespace std;
using namespace org::xmlBlaster::util;
using namespace org::xmlBlaster::client;
using namespace org::xmlBlaster;

class HelloWorld2 :public I_Callback
{
private:
   string  ME;
   Global& global_;
   Log&    log_;

public:
   HelloWorld2(Global& glob) : ME("HelloWorld2"), global_(glob), log_(glob.getLog())
   {
   }

   virtual ~HelloWorld2()
   {
   }

   void execute()
   {
      try {
         XmlBlasterAccess con(global_);
         log_.info(ME, "connecting to xmlBlaster");

         ConnectQos qos(global_, "joe", "secret");
         ConnectReturnQos retQos = con.connect(qos, this);  // Login to xmlBlaster, register for updates
         log_.info(ME, "successfully connected to xmlBlaster. Return qos: " + retQos.toXml());
         log_.info(ME, "subscribing to xmlBlaster");
         con.subscribe("<key oid='HelloWorld2'/>", "<qos/>");
         log_.info(ME, "successfully subscribed to xmlBlaster");

         PublishQos publishQos(global_);
         MessageUnit msgUnit(string("<key oid='HelloWorld2'/>"),
                                     string("Hi"),
                                     publishQos);

         log_.info(ME, "publishing to xmlBlaster");
         con.publish(msgUnit);
         log_.info(ME, "successfully published to xmlBlaster");
         try {
            TimestampFactory::sleepSecs(1);
         }
         catch(XmlBlasterException e) {
            cout << e.toXml() << endl;
         }

         con.erase("<key oid='HelloWorld2'/>", "<qos/>");
         con.disconnect("<qos/>");
      }
      catch (XmlBlasterException e) {
         cout << e.toXml() << endl;
      }
   }

   string update(const string& sessionId, UpdateKey& updateKey, void *content, long contentSize, UpdateQos& updateQos)
   {
      log_.info(ME, "update: unique Key: " + updateKey.getUniqueKey());
      log_.info(ME, "update: state     : " + updateQos.getState());
      log_.info(ME, "update: updateQos : " + updateQos.toXml());
      return "";
   }

};

/**
 * Try
 * <pre>
 *   java HelloWorld2 -help
 * </pre>
 * for usage help
 */
int main(int args, char ** argv)
{
   XMLPlatformUtils::Initialize();
   Global& glob = Global::getInstance();
   glob.initialize(args, argv);
// XmlBlasterConnection::usage();
//   glob.getLog().info("HelloWorld2", "Example: java HelloWorld2\n");

   HelloWorld2 hello(glob);
   hello.execute();
   return 0;
}
